src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/feature/AbstractUmapFeatureComputer.java

```diff
@@ -0,0 +1,239 @@
+ *
+ * @param <V> the type of vertex
+ * @param <G> the type of read-only graph
+ */
+public abstract class AbstractUmapFeatureComputer< V extends Vertex< ? >, G extends ReadOnlyGraph< V, ? > >
+		extends AbstractSerialFeatureComputer< V >
+{
+
+	private static final Logger logger = LoggerFactory.getLogger( MethodHandles.lookup().lookupClass() );
+
+	private final StatusService statusService;
+
+	private final Umap umap;
+
+	private List< UmapInputDimension< V > > inputDimensions;
+
+	private UmapFeatureSettings settings;
+
+	private AbstractUmapFeature< V > feature;
+
+	private double[][] umapResult;
+
+	private final IntPropertyMap< V > validFeaturesCache;
+
+	private int index;
+
+	protected AbstractUmapFeatureComputer( final Model model, final Context context )
+	{
+		this.model = model;
+		this.statusService = context.getService( StatusService.class );
+		this.umap = new Umap();
+		this.validFeaturesCache = createFeatureCacheInstance();
+	}
+
+	/**
+	 * Computes the UMAP feature with the given settings and input dimensions and declares it in the feature model.
+	 * <br>
+	 * During computation, the given graph is locked for reading.
+	 * The UMAP feature is computed for each vertex in the graph, excluding vertices with invalid data rows
+	 * (i.e. rows where the selected feature projections do not have values, such as {@link Double#NaN} or {@link Double#POSITIVE_INFINITY}).
+	 *
+	 * @param settings        the UMAP settings
+	 * @param inputDimensions the input dimensions
+	 * @param graph           the read-only graph
+	 */
+	public void computeFeature( final UmapFeatureSettings settings, final List< UmapInputDimension< V > > inputDimensions,
+			final G graph )
+	{
+		logger.info( "Computing UmapFeature with settings: {}", settings );
+		this.settings = settings;
+		logger.info( "Computing UmapFeatureComputer with {} input dimensions.", inputDimensions.size() );
+		for ( UmapInputDimension< V > inputDimension : inputDimensions )
+			logger.info( "Input dimension: {}", inputDimension );
+		this.inputDimensions = inputDimensions;
+		this.forceComputeAll = new AtomicBoolean( true );
+		long start = System.currentTimeMillis();
+		ReentrantReadWriteLock.ReadLock lock = getLock( graph ).readLock();
+		lock.lock();
+		try
+		{
+			run();
+		}
+		finally
+		{
+			lock.unlock();
+		}
+		logger.info( "Finished computing UmapFeature in {} ms", System.currentTimeMillis() - start );
+		model.getFeatureModel().declareFeature( feature );
+	}
+
+	@Override
+	protected void compute( final V vertex )
+	{
+		if ( isVertexValid( vertex ) )
+		{
+			for ( int i = 0; i < settings.getNumberOfOutputDimensions(); i++ )
+			{
+				DoublePropertyMap< V > umapOutput = feature.getUmapOutputMaps().get( i );
+				umapOutput.set( vertex, umapResult[ index ][ i ] );
+			}
+			index++;
```

Comment:
The usage of this "index" field in the class is horribly error prone. PLEASE replace it with a vertexToRowId map!!!! (You can use `RefToIntMap` or something like that.)